src/methods/function.js

```diff
@@ -1060,6 +1060,62 @@ export function EvaluateStatements(
 
 }
 
+export function PartiallyEvaluateStatements(
+  body: Array<BabelNodeStatement>, blockValue: void | NormalCompletion | Value,
+  strictCode: boolean, blockEnv: LexicalEnvironment, realm: Realm
+): [Completion | Value, Array<BabelNodeStatement>] {
+  let statementAsts = [];
+  for (let node of body) {
+    if (node.type !== "FunctionDeclaration") {
+      let [res, nast, nio] = blockEnv.partiallyEvaluateCompletionDeref(node, strictCode);
+      for (let ioAst of nio) statementAsts.push(ioAst);
+      statementAsts.push((nast: any));
```

Comment:
If nast is just a residual expression resulting from constant folding then evaluating it alone will not preserve any I/O side effects, hence the additional statements are the "residual I/O" of the evaluation of node and need to happen before the evaluation of nast. 

Another way to think of this is that ioAst and nast together form an expression of the form "let ioAst in nast". There is no way to express this as a single Babel AST node, hence the somewhat awkward expression of the idea.